### PR TITLE
Add search engine result preview on product page

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Add a search engine result preview in the product page, SEO tab
| Type?         | new feature
| Category?     | BO
| BC breaks?    | Nope
| Deprecations? | Nope
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10046
| How to test?  | You should have a Google style search preview in the SEO tab. Modifying the inputs meta title & meta description must update the preview.

capture du 2018-08-13 14-25-15

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/9444)

### CODE REVIEW

File: admin-dev/themes/new-theme/js/app/utils/serp/index.js [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+ * needs please refer to http://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2018 PrestaShop SA
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+import Vue from 'vue';
+import serp from './serp';
+
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/serp/index.js]: `const $ = window.$;` is missing?

File: admin-dev/themes/new-theme/js/app/utils/serp/serp.vue [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+<script>
+  export default {
+    name: 'serp',
+    data: () => ({
+        title: '',
+        url: '',
+        description: '',
+        defaultUrl: 'http://example.com/',
+    }),
+    methods: {
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/serp/serp.vue]: You can use `computed` instead methods to use a var for display :)

File: admin-dev/themes/new-theme/js/app/utils/serp/index.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        this.defaultTitle = $('.serp-default-title:input');
+        this.watchedTitle = $('.serp-watched-title:input');
+        this.defaultDescription = $('.serp-default-description:input');
+        this.watchedDescription = $('.serp-watched-description:input');
+        this.defaultUrl = $('.serp-default-url:input');
+
+        this.vm = new Vue({
+          el: '#serp-app',
+          template: '<serp ref="serp" />',
+          components: { serp },
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/serp/index.js]: Should be better to use `props` for component to make sure it's reusable.

File: admin-dev/themes/new-theme/js/app/utils/serp/serp.vue [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+        setTitle(title) {
```
Comment (PierreRambaud) [inline admin-dev/themes/new-theme/js/app/utils/serp/serp.vue]: If you use props, you can have an easier usage with `computed`

File: admin-dev/themes/new-theme/js/app/utils/serp/serp.vue [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            if(description.length > 150) {
+                description = description.substring(0, 150) + ' ...';
+            }
+            this.description = description;
+        }
+    }
+  };
+</script>
+
+<style lang="sass" type="text/scss">
```
Comment (mickaelandrieu) [inline admin-dev/themes/new-theme/js/app/utils/serp/serp.vue]: Not sure it's a good or bad idea, but you can import a file instead:

```html
<style lang="scss">
@import 'where/you/want/serp.css';
</style>
```

File: src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/Forms/form_seo.html.twig [hunk 6]
```diff
@@ -26,6 +26,9 @@
 
   <h2>{{ 'Search Engine Optimization'|trans({}, 'Admin.Catalog.Feature') }}</h2>
   <p class="subtitle">{{ 'Improve your ranking and how your product page will appear in search engines results.'|trans({}, 'Admin.Catalog.Feature') }}</p>
+  
+  <p>{{ "Search engine result preview:"|trans({}, 'Admin.Catalog.Feature') }}</p>
+  <div id="serp-app"></div>
```
Comment (mickaelandrieu) [inline src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/Forms/form_seo.html.twig]: I guess this needs a little bit docs: how a module can alter this block?

```diff html
- <div id="serp-app"></div>
+ {% block product_seo_serp %} // take a look at current names of block, I don't remember the convention
+     <!-- @see link to the serp component file -->
+     <div id="serp-app"></div>
+ {% endblock %}
```

File: admin-dev/themes/new-theme/js/app/utils/serp/index.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+ * 
+ * The component is automatically updated by watching several inputs.
+ * Set the proper class to link a input to a part of the panel.
+ * 
+ * @returns {serpUtil.indexAnonym$0}
+ */
+class SerpApp {
+  constructor() {
+    // Stop if ID not found
+    if (0 === $("#serp-app").length) {
```
Comment (mickaelandrieu) [inline admin-dev/themes/new-theme/js/app/utils/serp/index.js]: I'm curious about it: as this file is only loaded in product page, in which case do you need to disable it?

Also, I'll prefer a better comment like: `Disable application If <serp /> markup is not available`

**Review by mickaelandrieu — COMMENTED** [review]
few comments here, I'm anxious about adding Vuejs in such broken page lol

File: admin-dev/themes/new-theme/js/app/utils/serp/serp.vue (hunk 5 above)
Comment (PierreRambaud) [reply admin-dev/themes/new-theme/js/app/utils/serp/serp.vue]: VueJs style guide https://vuejs.org/v2/style-guide/#Component-style-scoping-essential

File: admin-dev/themes/new-theme/js/app/utils/serp/index.js (hunk 7 above)
Comment (Quetzacoalt91) [reply admin-dev/themes/new-theme/js/app/utils/serp/index.js]: It's written in the doc before the class

Comment (Quetzacoalt91) [reply admin-dev/themes/new-theme/js/app/utils/serp/index.js]: And later, this component will be added in the SEO & URLs, then in the category page as soon as they are migrated.

File: admin-dev/themes/new-theme/js/app/utils/serp/serp.vue (hunk 5 above)
Comment (mickaelandrieu) [reply admin-dev/themes/new-theme/js/app/utils/serp/serp.vue]: and? I'm not saying we should change the scope here, I'm saying we may put scss file in the right location (easier to change for people who are not aware at all to Vuejs)

Comment (PierreRambaud) [reply admin-dev/themes/new-theme/js/app/utils/serp/serp.vue]: it's just easier to play with component and keep specific style in its context ^^

Comment (mickaelandrieu) [reply admin-dev/themes/new-theme/js/app/utils/serp/serp.vue]: maybe, I don't know and it's certainly not me who will try to customize the design of our back office :)

File: src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/Forms/form_seo.html.twig [hunk 8]
```diff
@@ -26,6 +26,9 @@
 
   <h2>{{ 'Search Engine Optimization'|trans({}, 'Admin.Catalog.Feature') }}</h2>
   <p class="subtitle">{{ 'Improve your ranking and how your product page will appear in search engines results.'|trans({}, 'Admin.Catalog.Feature') }}</p>
+  
+  <p>{{ "Search engine result preview:"|trans({}, 'Admin.Catalog.Feature') }}</p>
```
Comment (LouiseBonnard) [inline src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/Forms/form_seo.html.twig]: Wouldn't it be nice to have a more user-friendly phrase here? Like "Get a sneak peek at it:", "Have a look at the result:", or "Here is a preview of your search engine result, play with it!".

Comment (PierreRambaud) [reply src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/Forms/form_seo.html.twig]: I like the last one `Here is a preview of your search engine result, play with it!` :+1:

Comment (Quetzacoalt91) [reply src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/Forms/form_seo.html.twig]: Indeed, I requested a better phrase on Jira, the current one needs to be replaced. :)

ping @colinegin

Comment (colinegin) [reply src/PrestaShopBundle/Resources/views/Admin/Product/ProductPage/Forms/form_seo.html.twig]: Love it !

**Comment by Quetzacoalt91** [comment]
@mickaelandrieu, I understand that you're concerned about loading new feature to the product page, but note this component and the global page behavior are not dependent from each other.

This component will be also added on two other pages, once migrated:
* SEO & URLs
* Category

**Comment by PierreRambaud** [comment]
This one must be rebased please @Quetzacoalt91  :)

**Review by mickaelandrieu — DISMISSED** [review]

**Comment by marionf** [comment]
@Quetzacoalt91 

If I edit the meta description, it's not changed

capture d ecran_279

**Comment by Quetzacoalt91** [comment]
See https://github.com/PrestaShop/PrestaShop/pull/9444/commits/6c6764117d295d6a889d1b4125d7aa92f5b16245 for the fix.

**Review by matks — DISMISSED** [review]

**Review by matks — APPROVED** [review]

**Comment by matks** [comment]
@marionf You validated this PR but it got back into "To be review" state. I think the changes should not modify your validation however I leave the decision to re-validate or allow the merge to you.

**Comment by marionf** [comment]
I did a last check and it's all good, so let's merge it :)

**Pull request merged**